### DISCUSSION TITLE
OAuth Simulation

### DISCUSSION THREAD

**Discussion by bigste** [mockoon/mockoon#1089] [General]
I want to simulate an OAuth style API, where...

- The inital call sends a username:password combo, 
- Mockoon responds with a randomly generated AccessToken, 
- All subsequent requests must use the AccessToken in the header 
- The AccessToken can be verified in Mockoon because I've stored it in a Data Bucket, or a variable or something?

I can't see how I can achieve this with the Helper tools available. I could do it with a fixed AccessToken (from a data bucket), or I could probably generate a fresh (but fixed) AccessToken in the dtaa bucket and use that. But a RefreshToken would need to be generated inline and saved for other Routes.

Or have I missed the whole point of this and I'm just describing an actual API server and this isn't that, it's a Mock API server?

**Comment by 255kb** [comment]
> Or have I missed the whole point of this and I'm just describing an actual API server and this isn't that, it's a Mock API server?

I would say, many things could be included in Mockoon to make it more realistic. But at one point, if it's too much config, it's maybe better to rely on the real "thing". The question is where to put the limit. 

What you describe can be "easily" achieved by just hardcoding some token in a data bucket. and returning it in a login route. The only missing part is the ability to reference a data bucket with templating helpers inside the rules. We can already do this in the headers that can be dynamically generated, but not the rules. There is an [opened issue](https://github.com/mockoon/mockoon/issues/988) that could be done quickly. It's a logical evolution. 
After this is done you could just put a fake token `abcdxxxxx` inside a bucket and use it to serve the login route body and add a rule for responses by putting `{{data 'mytoken'}}` in the rule value. 
After that, I am not sure generating realistic tokens and refresh tokens would add a lot of value. 
I would see more value to being able to define global response and rules to apply this kind of check an all the routes.
